ERCS/erc-8006.md

```diff
@@ -0,0 +1,370 @@
+```
+
+### Interfaces
+
+#### Artifact Interfaces
+
+Artifacts may implement any logic but MUST adhere to a standard interface that allows the policy controller to integrate them correctly and ensure consistent data flow:
+
+```solidity
+interface IArbitraryDataArtifact {
+    function exec(bytes[] memory data) external returns (bytes memory);
+
+    function init(bytes memory data) external;
+
+    function getExecDescriptor()
+        external
+        pure
+        returns (string[] memory argsNames, string[] memory argsTypes, string memory returnType);
+
+    function getInitDescriptor()
+        external
+        pure
+        returns (string[] memory argsNames, string[] memory argsTypes);
+
+    function description() external pure returns (string memory desc);
+}
+```
+
+
+Other interfaces are more implicit, following conventions and approaches explained in the Rationale section.
+
+### Policy Handler
+
+The Policy Handler orchestrates artifact execution. While implementations may vary, this standard RECOMMENDS a **Directed Acyclic Graph (DAG)** as the reference architecture (see [reference implementation](../assets/erc-8006/contracts/inheritance/DAGWithPolicyMetadata.sol) and [Policy Handler](../assets/erc-8006/contracts/PolicyHandler.sol)).
+
+During **policy initialization**, the handler: 
+1. Constructs a DAG from the specified artifacts and their configurations
+2. Receives the **Root** node identifier (referencing the starting artifact instance in the policy)
+3. Each stateful artifact instance is initialized as needed during this phase, ensuring proper setup before policy evaluation
+
+When **policy evaluation** is requested, the handler:
+1. Collects the necessary variables for evaluation
+2. Initiates a recursive *traversal* of the graph
+
+For each node during the **traversal**, the handler:
+1. Calls the target artifact using the standard interface
+2. Supplies it with the appropriate variables and/or results from previously evaluated (calculated) nodes that the current node depends on
+3. Collects the output values of the artifacts for use by parent nodes
+4. Finishes the traversal once all nodes have been visited and evaluated
+
+In summary, the handler passes outputs from child nodes as inputs to parent nodes; artifacts do not invoke each other directly. After the traversal completes, the root node’s output value represents the policy’s final result. This design provides a structured mechanism for rule interaction, addressing the challenge of composing complex rules from simpler components.
+
+![Policy bird's eye view](../assets/erc-8006/simple_policy_bird_eye_view.svg)
+
+The handler SHOULD provide a method to retrieve information about all variables used by all artifacts in the policy:
+
+```solidity
+function getVariablesList() public view returns (ExecVarsMetadata[] memory list) {
+    require(isInitialized, POLICY_NOT_INITIALIZED_ERR);
+
+    list = Utils.getVarsDesriptionList(dag.getNodes());
```

Comment:
This is an implementation detail. ERCs should only specify the externally visible behaviour of a system, not how that system is built.